File: golangexamples/leaky_abstractions.go
package main

/*
ANTI-PATTERN: Leaky Abstractions

"All non-trivial abstractions, to some degree, are leaky." - Joel Spolsky

An abstraction that claims to hide implementation details but forces you to understand
the underlying implementation to use it correctly or efficiently.
*/

import (
	"context"
	"errors"
	"fmt"
	"io"
	"strings"
	"sync"
	"time"
)

// ==============================================================================
// Example 1: Interface Leaking Through Error Types
// ==============================================================================

// Cache interface claims to abstract storage, but errors leak the implementation
type Cache interface {
	Get(key string) (string, error)
	Set(key string, value string) error
	Delete(key string) error
}

// RedisCache implementation
type RedisCache struct {
	connected bool
}

// LEAK: Redis-specific errors
var (
	ErrRedisPoolExhausted = errors.New("redis: connection pool exhausted")
	ErrRedisTimeout       = errors.New("redis: operation timeout")
	ErrRedisConnectionLost = errors.New("redis: connection lost")
)

func (r *RedisCache) Get(key string) (string, error) {
	if !r.connected {
		return "", ErrRedisConnectionLost // LEAK: Implementation-specific error
	}
	return "value", nil
}

func (r *RedisCache) Set(key string, value string) error {
	if !r.connected {
		return ErrRedisPoolExhausted // LEAK: Implementation-specific error
	}
	return nil
}

func (r *RedisCache) Delete(key string) error {
	return nil
}

// LEAKY ABSTRACTION: Code that uses Cache is coupled to Redis implementation
func ProcessDataBAD(cache Cache) error {
	/*
		PROBLEM: The interface claims to hide the implementation, but we're
		forced to check for Redis-specific errors!

		You must understand:
		- Redis connection pooling
		- Redis-specific error conditions
		- Implementation details that should be hidden

		The abstraction leaks through error types.
	*/
	err := cache.Set("key", "value")

	// LEAK: Checking for implementation-specific error!
	if err == ErrRedisPoolExhausted {
		// Now we're coupled to Redis - the abstraction has leaked
		fmt.Println("Redis pool exhausted - retry logic")
		time.Sleep(100 * time.Millisecond)
		return cache.Set("key", "value")
	}

	// LEAK: Another Redis-specific error check
	if err == ErrRedisTimeout {
		fmt.Println("Redis timeout - adjust timeout settings")
	}

	return err
}

// Better approach: Use error wrapping to hide implementation
type CacheError struct {
	Op  string
	Err error
}

func (e *CacheError) Error() string {
	return fmt.Sprintf("cache %s: %v", e.Op, e.Err)
}

func (e *CacheError) Unwrap() error {
	return e.Err
}

// ==============================================================================
// Example 2: io.Reader/Writer Leaking Buffer Management
// ==============================================================================

// DataReader claims to abstract data reading, but leaks buffer behavior
type DataReader struct {
	source io.Reader
	buffer []byte
}

func NewDataReader(source io.Reader, bufferSize int) *DataReader {
	return &DataReader{
		source: source,
		buffer: make([]byte, bufferSize),
	}
}

func (dr *DataReader) ReadNext() (string, error) {
	/*
		LEAK: The abstraction looks simple, but you need to understand:
		- Buffer size affects performance dramatically
		- EOF handling semantics from io.Reader
		- When buffers are allocated vs reused
		- Memory implications of buffer size

		The abstraction leaks io.Reader semantics and buffer management.
	*/
	n, err := dr.source.Read(dr.buffer)
	if err != nil && err != io.EOF {
		return "", err
	}
	return string(dr.buffer[:n]), nil
}

func ProcessFileBAD() {
	/*
		PROBLEM: This looks simple, but performance completely depends on
		understanding buffer management which should be hidden!
	*/
	reader := NewDataReader(strings.NewReader("hello world"), 1)

	// LEAK: Tiny buffer size causes many system calls
	// You need to understand this to use the abstraction efficiently!
	data, _ := reader.ReadNext()
	fmt.Println(data)
}

func ProcessFileGOOD() {
	/*
		SOLUTION: Use appropriate buffer size, but this means understanding
		the underlying I/O model. The abstraction leaked!
	*/
	reader := NewDataReader(strings.NewReader("hello world"), 4096)

	data, _ := reader.ReadNext()
	fmt.Println(data)
}

// ==============================================================================
// Example 3: sync.Pool Leaking Memory Semantics
// ==============================================================================

// ObjectPool claims to manage object lifecycle, but leaks GC behavior
type ObjectPool struct {
	pool *sync.Pool
}

func NewObjectPool() *ObjectPool {
	return &ObjectPool{
		pool: &sync.Pool{
			New: func() interface{} {
				return &ExpensiveObject{}
			},
		},
	}
}

type ExpensiveObject struct {
	Data [1024]byte
}

func (op *ObjectPool) Get() *ExpensiveObject {
	/*
		LEAK: This looks like simple object pooling, but you must understand:
		- sync.Pool can evict objects at any GC
		- Objects are not guaranteed to be available
		- Memory semantics differ from manual pooling
		- No size limits - can grow unbounded

		The abstraction leaks Go's GC behavior and sync.Pool semantics.
	*/
	return op.pool.Get().(*ExpensiveObject)
}

func (op *ObjectPool) Put(obj *ExpensiveObject) {
	op.pool.Put(obj)
}

func UsePoolBAD() {
	/*
		PROBLEM: This looks like guaranteed object reuse, but sync.Pool
		doesn't guarantee anything! GC can clear it at any time.
	*/
	pool := NewObjectPool()

	obj := pool.Get()
	// Use object
	pool.Put(obj)

	// LEAK: You assume Get() returns the same object, but it might not!
	// You need to understand sync.Pool behavior to use it correctly.
	obj2 := pool.Get()
	if obj2 != obj {
		fmt.Println("Different object - sync.Pool was cleared!")
	}
}

// ==============================================================================
// Example 4: http.Client Leaking Connection Management
// ==============================================================================

// HTTPClient claims to abstract HTTP, but leaks connection pooling
type HTTPClient struct {
	// LEAK: These fields leak implementation details users need to understand
	MaxIdleConns        int
	MaxIdleConnsPerHost int
	IdleConnTimeout     time.Duration
}

func NewHTTPClient() *HTTPClient {
	return &HTTPClient{
		MaxIdleConns:        100,
		MaxIdleConnsPerHost: 2,
		IdleConnTimeout:     90 * time.Second,
	}
}

func (hc *HTTPClient) Get(url string) (string, error) {
	/*
		LEAK: Simple Get() hides complex connection pooling behavior.

		You must understand:
		- Connection pooling and reuse
		- Keep-alive behavior
		- DNS caching
		- TLS session resumption
		- When connections are closed

		The abstraction leaks TCP connection management.
	*/
	fmt.Printf("[HTTP] GET %s (using connection pool)\n", url)
	return "response", nil
}

func MakeRequestsBAD() {
	/*
		PROBLEM: Looks simple, but connection behavior is completely opaque.
		Performance depends on understanding connection pooling!
	*/
	client := NewHTTPClient()

	// LEAK: Are these using the same connection? New connections?
	// You can't tell without understanding the implementation!
	for i := 0; i < 10; i++ {
		_, _ = client.Get(fmt.Sprintf("http://api.example.com/users/%d", i))
	}

	// LEAK: What happens to idle connections? When are they closed?
	// You need to understand MaxIdleConnsPerHost to tune performance!
}

// ==============================================================================
// Example 5: Channel Buffering Leaking Goroutine Behavior
// ==============================================================================

// WorkQueue claims to abstract async work, but leaks channel semantics
type WorkQueue struct {
	tasks chan func()
}

func NewWorkQueue(bufferSize int) *WorkQueue {
	/*
		LEAK: Buffer size dramatically affects behavior, but this is
		supposed to be an implementation detail!

		You must understand:
		- Buffered vs unbuffered channels
		- Blocking behavior
		- Goroutine scheduling
		- Memory usage of buffered channels
	*/
	return &WorkQueue{
		tasks: make(chan func(), bufferSize),
	}
}

func (wq *WorkQueue) Submit(task func()) error {
	/*
		LEAK: This looks like it always succeeds, but it blocks!
		You need to understand channel semantics to use it correctly.
	*/
	select {
	case wq.tasks <- task:
		return nil
	default:
		return fmt.Errorf("queue full")
	}
}

func (wq *WorkQueue) Start() {
	go func() {
		for task := range wq.tasks {
			task()
		}
	}()
}

func UseWorkQueueBAD() {
	/*
		PROBLEM: Behavior completely changes based on buffer size,
		but this is supposed to be hidden by the abstraction!
	*/
	queue := NewWorkQueue(1) // Tiny buffer
	queue.Start()

	// LEAK: This works fine
	_ = queue.Submit(func() { fmt.Println("Task 1") })

	// LEAK: This might fail because buffer is full!
	// You need to understand channel buffering to use this correctly.
	err := queue.Submit(func() { fmt.Println("Task 2") })
	if err != nil {
		fmt.Println("Queue full - needed to understand buffering!")
	}
}

func UseWorkQueueGOOD() {
	/*
		SOLUTION: Use larger buffer, but this means understanding
		the channel implementation. The abstraction leaked!
	*/
	queue := NewWorkQueue(1000)
	queue.Start()

	for i := 0; i < 100; i++ {
		_ = queue.Submit(func() { fmt.Println("Task") })
	}
}

// ==============================================================================
// Example 6: context.Context Leaking Cancellation Propagation
// ==============================================================================

// Service claims to abstract business logic, but leaks context semantics
type Service struct {
	name string
}

func (s *Service) ProcessRequest(ctx context.Context, request string) error {
	/*
		LEAK: This looks simple, but you must understand:
		- Context cancellation propagation
		- Deadline vs timeout vs cancellation
		- When to check context.Done()
		- Context value propagation

		The abstraction leaks context cancellation semantics.
	*/
	select {
	case <-ctx.Done():
		// LEAK: You need to check this, but when? How often?
		// The abstraction doesn't tell you!
		return ctx.Err()
	case <-time.After(100 * time.Millisecond):
		fmt.Printf("[%s] Processed: %s\n", s.name, request)
		return nil
	}
}

func CallServiceBAD() {
	/*
		PROBLEM: Context cancellation behavior is opaque.
		You need to understand context semantics to use this correctly!
	*/
	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
	defer cancel()

	service := &Service{name: "UserService"}

	// LEAK: This will timeout, but the error tells you nothing about
	// whether it was context timeout, cancellation, or deadline!
	err := service.ProcessRequest(ctx, "data")
	if err != nil {
		// LEAK: Is this context.DeadlineExceeded or context.Canceled?
		// You need to understand context error types!
		fmt.Printf("Error: %v\n", err)
	}
}

// ==============================================================================
// Example 7: Database Abstraction Leaking Transaction Semantics
// ==============================================================================

// DB claims to abstract database, but leaks transaction behavior
type DB struct {
	inTransaction bool
}

func (db *DB) Query(query string) ([]map[string]interface{}, error) {
	/*
		LEAK: This looks simple, but you must understand:
		- Transaction isolation levels
		- Lock behavior
		- Rollback semantics
		- Connection pooling

		The abstraction leaks SQL transaction semantics.
	*/
	fmt.Printf("[DB] Query: %s (in_transaction: %v)\n", query, db.inTransaction)
	return nil, nil
}

func (db *DB) Begin() error {
	/*
		LEAK: What isolation level? What happens to concurrent queries?
		You need to understand transaction semantics!
	*/
	db.inTransaction = true
	return nil
}

func (db *DB) Commit() error {
	db.inTransaction = false
	return nil
}

func (db *DB) Rollback() error {
	db.inTransaction = false
	return nil
}

func UseDBBAD() {
	/*
		PROBLEM: Transaction behavior is opaque. What happens if you
		forget to commit? What isolation level? The abstraction leaks!
	*/
	db := &DB{}

	_ = db.Begin()
	_, _ = db.Query("UPDATE users SET balance = balance - 100 WHERE id = 1")

	// LEAK: What happens if we crash here? Auto-rollback? Committed?
	// You need to understand transaction semantics!

	_ = db.Commit()
}

// ==============================================================================
// MAIN - Demonstrate Leaky Abstractions
// ==============================================================================

func main() {
	fmt.Println("=" + strings.Repeat("=", 79))
	fmt.Println("LEAKY ABSTRACTIONS ANTI-PATTERN")
	fmt.Println("=" + strings.Repeat("=", 79))
	fmt.Println()

	fmt.Println("Example 1: Interface Leaking Through Error Types")
	fmt.Println(strings.Repeat("-", 80))
	cache := &RedisCache{connected: false}
	_ = ProcessDataBAD(cache)
	fmt.Println()

	fmt.Println("Example 2: io.Reader Leaking Buffer Management")
	fmt.Println(strings.Repeat("-", 80))
	ProcessFileBAD()
	fmt.Println()

	fmt.Println("Example 3: sync.Pool Leaking Memory Semantics")
	fmt.Println(strings.Repeat("-", 80))
	UsePoolBAD()
	fmt.Println()

	fmt.Println("Example 5: Channel Buffering Leaking Goroutine Behavior")
	fmt.Println(strings.Repeat("-", 80))
	UseWorkQueueBAD()
	time.Sleep(200 * time.Millisecond) // Wait for goroutines
	fmt.Println()

	fmt.Println("Example 6: context.Context Leaking Cancellation")
	fmt.Println(strings.Repeat("-", 80))
	CallServiceBAD()
	fmt.Println()

	fmt.Println("Example 7: Database Leaking Transaction Semantics")
	fmt.Println(strings.Repeat("-", 80))
	UseDBBAD()
	fmt.Println()

	fmt.Println("=" + strings.Repeat("=", 79))
	fmt.Println("KEY INSIGHT: All abstractions leak to some degree.")
	fmt.Println("Good abstractions leak gracefully and document what leaks.")
	fmt.Println("Bad ones force you to understand implementation details to use")
	fmt.Println("them correctly, defeating the purpose of abstraction.")
	fmt.Println("=" + strings.Repeat("=", 79))
}

File: pythonexamples/leaky_abstractions.py
"""
ANTI-PATTERN: Leaky Abstractions

"All non-trivial abstractions, to some degree, are leaky." - Joel Spolsky

An abstraction that claims to hide implementation details but forces you to understand
the underlying implementation to use it correctly or efficiently.
"""

# ==============================================================================
# Example 1: ORM - The Classic Leaky Abstraction
# ==============================================================================

class User:
    """Simulated ORM model"""
    def __init__(self, id, name, company_id):
        self.id = id
        self.name = name
        self.company_id = company_id
        self._profile = None

    @property
    def profile(self):
        if self._profile is None:
            # LEAK: This looks like attribute access but triggers a database query!
            print(f"[DB QUERY] SELECT * FROM profiles WHERE user_id = {self.id}")
            self._profile = Profile(self.id)
        return self._profile


class Profile:
    """Simulated profile model"""
    def __init__(self, user_id):
        self.user_id = user_id
        self.avatar_url = f"https://cdn.example.com/avatar_{user_id}.png"
        self.bio = "Sample bio"


class QuerySet:
    """Simulated ORM QuerySet"""
    def __init__(self, data):
        self.data = data
        self._prefetched = False

    def filter(self, **kwargs):
        print(f"[DB QUERY] SELECT * FROM users WHERE {kwargs}")
        return self

    def select_related(self, *fields):
        """
        LEAK: You need to understand SQL JOINs to use the ORM efficiently!
        The abstraction claims to hide SQL but you must know about JOINs.
        """
        print(f"[DB QUERY] SELECT * FROM users LEFT JOIN {fields}")
        self._prefetched = True
        return self

    def all(self):
        return self.data


# LEAKY ABSTRACTION: The ORM looks simple but hides performance traps
def get_users_with_avatars_BAD(company_id):
    """
    PROBLEM: This looks clean but causes N+1 query problem!

    You must understand:
    - SQL query execution
    - Lazy loading behavior
    - Database round-trip costs

    The abstraction leaks because simple Python code causes hidden database queries.
    """
    users = QuerySet([
        User(1, "Alice", company_id),
        User(2, "Bob", company_id),
        User(3, "Charlie", company_id)
    ]).filter(company_id=company_id).all()

    avatars = []
    for user in users:
        # LEAK: Each access triggers a hidden database query!
        # Looks like simple attribute access, but it's not.
        avatars.append(user.profile.avatar_url)

    return avatars


def get_users_with_avatars_GOOD(company_id):
    """
    SOLUTION: You must understand SQL to use the ORM correctly.
    The abstraction has leaked - you need implementation knowledge.
    """
    users = QuerySet([
        User(1, "Alice", company_id),
        User(2, "Bob", company_id),
        User(3, "Charlie", company_id)
    ]).filter(company_id=company_id).select_related('profile').all()

    avatars = []
    for user in users:
        avatars.append(user.profile.avatar_url)

    return avatars


# ==============================================================================
# Example 2: File Context Manager - Leaks Buffer Behavior
# ==============================================================================

class BufferedFileReader:
    """
    LEAKY ABSTRACTION: Claims to hide file I/O complexity but leaks buffering behavior
    """
    def __init__(self, filename, buffer_size=8192):
        self.filename = filename
        self.buffer_size = buffer_size
        self.file = None

    def __enter__(self):
        self.file = open(self.filename, 'r')
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.file:
            self.file.close()

    def read_line(self):
        """
        LEAK: This looks like it reads one line, but the buffering behavior
        means it actually reads buffer_size bytes from disk!

        Users must understand:
        - OS-level buffering
        - Filesystem caching
        - Memory vs disk I/O tradeoffs
        """
        return self.file.readline()


def process_large_file_BAD():
    """
    PROBLEM: The abstraction leaks - you need to understand buffering
    to know why this might be slow or memory-intensive.
    """
    with BufferedFileReader('/tmp/large_file.txt') as reader:
        line = reader.read_line()
        # LEAK: Even though you only read one line, the entire buffer
        # was loaded. This is invisible to the user!
        return line


def process_large_file_GOOD():
    """
    SOLUTION: You must understand the buffering to use it efficiently.
    The abstraction leaked - you need implementation knowledge.
    """
    # You need to know about buffer_size to tune performance
    with BufferedFileReader('/tmp/large_file.txt', buffer_size=64) as reader:
        line = reader.read_line()
        return line


# ==============================================================================
# Example 3: Thread Pool Executor - Leaks Threading Model
# ==============================================================================

from concurrent.futures import ThreadPoolExecutor
import time

def expensive_io_task(task_id):
    """Simulates I/O-bound work"""
    time.sleep(0.1)
    return f"Task {task_id} complete"


def cpu_intensive_task(n):
    """
    LEAKY ABSTRACTION: ThreadPoolExecutor looks great for parallelism,
    but leaks the GIL (Global Interpreter Lock) limitation!

    LEAK: You must understand:
    - Python's GIL prevents true parallel CPU work
    - Thread vs Process differences
    - When to use ThreadPoolExecutor vs ProcessPoolExecutor
    """
    total = 0
    for i in range(n):
        total += i ** 2
    return total


def parallel_work_BAD():
    """
    PROBLEM: This looks like it will run in parallel, but the GIL means
    CPU-bound tasks won't actually benefit from multiple threads!

    The abstraction leaks - you need to understand the GIL.
    """
    with ThreadPoolExecutor(max_workers=4) as executor:
        # LEAK: These look like they'll run in parallel, but GIL prevents it!
        futures = [executor.submit(cpu_intensive_task, 1000000) for _ in range(4)]
        results = [f.result() for f in futures]
    return results


def parallel_work_GOOD():
    """
    SOLUTION: For I/O-bound work, threads work great.
    But you needed to understand the GIL to know the difference!
    """
    with ThreadPoolExecutor(max_workers=4) as executor:
        # These actually run in parallel because they're I/O-bound
        futures = [executor.submit(expensive_io_task, i) for i in range(4)]
        results = [f.result() for f in futures]
    return results


# ==============================================================================
# Example 4: Requests Library - Leaks Connection Pooling
# ==============================================================================

class HTTPClient:
    """
    LEAKY ABSTRACTION: Simple HTTP client that leaks connection management
    """
    def __init__(self):
        self._connections = {}

    def get(self, url):
        """
        LEAK: This looks like a simple GET request, but there's hidden
        connection pooling, timeout behavior, and retry logic.

        Users must understand:
        - TCP connection lifecycle
        - Connection pooling and reuse
        - Timeout vs connection timeout vs read timeout
        - DNS caching
        """
        print(f"[HTTP] GET {url}")
        # Simulated connection pooling (hidden from user)
        host = url.split('/')[2]
        if host not in self._connections:
            print(f"[TCP] Opening new connection to {host}")
            self._connections[host] = "connection_object"
        else:
            print(f"[TCP] Reusing existing connection to {host}")

        return {"status": 200, "data": "response"}


def make_requests_BAD():
    """
    PROBLEM: Looks simple, but performance depends on understanding
    connection pooling, which is hidden by the abstraction.
    """
    client = HTTPClient()

    # LEAK: Each call reuses connections, but you can't see it!
    # You need to understand this to reason about performance
    for i in range(10):
        response = client.get(f"https://api.example.com/users/{i}")

    # LEAK: What happens to connections when client goes out of scope?
    # Are they closed? Kept alive? You need to know!


def make_requests_GOOD():
    """
    SOLUTION: Use session explicitly, but this means understanding
    the underlying HTTP connection model. The abstraction leaked!
    """
    client = HTTPClient()

    # You understand connection reuse and manage it explicitly
    try:
        for i in range(10):
            response = client.get(f"https://api.example.com/users/{i}")
    finally:
        # Explicitly clean up (if the API provided this)
        pass


# ==============================================================================
# Example 5: List Comprehension vs Generator - Leaks Memory Model
# ==============================================================================

def process_large_dataset_BAD(data_size=1000000):
    """
    LEAKY ABSTRACTION: List comprehension looks like generator but isn't!

    LEAK: You must understand:
    - Memory allocation for lists
    - Eager vs lazy evaluation
    - Iterator protocol

    The abstraction leaks because similar syntax has wildly different behavior.
    """
    # LEAK: This creates entire list in memory immediately!
    # Looks innocent but could cause OOM for large datasets
    numbers = [x * 2 for x in range(data_size)]

    # Now you have 1 million integers in memory
    return sum(numbers)


def process_large_dataset_GOOD(data_size=1000000):
    """
    SOLUTION: Use generator, but you needed to understand memory model
    to know when to use which! The abstraction leaked.
    """
    # Generator expression - lazy evaluation
    numbers = (x * 2 for x in range(data_size))

    # Memory efficient - but you needed to know this existed!
    return sum(numbers)


# ==============================================================================
# Example 6: Property Decorator - Leaks Computation Cost
# ==============================================================================

class DataAnalyzer:
    """
    LEAKY ABSTRACTION: @property makes methods look like attributes,
    hiding expensive computation.
    """
    def __init__(self, data):
        self._data = data

    @property
    def average(self):
        """
        LEAK: Looks like simple attribute access, but it's expensive computation!

        Users must understand:
        - Properties can run arbitrary code
        - Each access recomputes (no caching)
        - Performance characteristics are hidden
        """
        print("[EXPENSIVE] Calculating average...")
        return sum(self._data) / len(self._data)

    @property
    def sorted_data(self):
        """
        LEAK: Even worse - this sorts on every access!
        Looks like an attribute but has O(n log n) cost.
        """
        print("[EXPENSIVE] Sorting data...")
        return sorted(self._data)


def analyze_data_BAD():
    """
    PROBLEM: Looks like simple attribute access, but triggers
    expensive computation multiple times!
    """
    analyzer = DataAnalyzer(list(range(10000)))

    # LEAK: Each of these looks like attribute access but is expensive!
    print(analyzer.average)  # O(n) computation
    print(analyzer.average)  # O(n) computation AGAIN!
    print(analyzer.sorted_data)  # O(n log n) computation
    print(analyzer.sorted_data)  # O(n log n) computation AGAIN!

    # The abstraction leaks - you need to know it's not cached!


def analyze_data_GOOD():
    """
    SOLUTION: Cache the results, but you needed to understand
    the property was recomputing! Abstraction leaked.
    """
    analyzer = DataAnalyzer(list(range(10000)))

    # Explicitly cache because you know it's expensive
    avg = analyzer.average  # Compute once
    sorted_data = analyzer.sorted_data  # Compute once

    # Use cached values
    print(avg)
    print(sorted_data)


# ==============================================================================
# MAIN - Demonstrate the Leaky Abstractions
# ==============================================================================

def main():
    print("=" * 80)
    print("LEAKY ABSTRACTIONS ANTI-PATTERN")
    print("=" * 80)
    print()

    print("Example 1: ORM Leaky Abstraction (N+1 Query Problem)")
    print("-" * 80)
    print("BAD: Hidden N+1 queries")
    get_users_with_avatars_BAD(1)
    print()
    print("GOOD: Explicit join (but requires SQL knowledge!)")
    get_users_with_avatars_GOOD(1)
    print()

    print("=" * 80)
    print("Example 4: HTTP Client Connection Pooling Leak")
    print("-" * 80)
    make_requests_BAD()
    print()

    print("=" * 80)
    print("Example 6: Property Decorator Computation Cost Leak")
    print("-" * 80)
    analyze_data_BAD()
    print()

    print("=" * 80)
    print("KEY INSIGHT: All abstractions leak to some degree.")
    print("Good abstractions leak gracefully. Bad ones force you to")
    print("understand the implementation to use them correctly.")
    print("=" * 80)


if __name__ == "__main__":
    main()
